libs/src/bigint.sw

```diff
@@ -0,0 +1,669 @@
+            result.push(sum);
+            carry = c;
+            idx += 1;
+        }
+
+        if carry > 0 {
+            result.push(carry);
+        }
+
+        // Remove leading zeros
+        while result.len() > 1 && result.last() == Some(0) {
+            let _ = result.pop();
+        }
+
+        Self { limbs: result }
+    }
+}
+
+fn mac_with_carry_u64(acc: u64, a: u64, b: u64, c: u64) -> (u64, u64) {
+    let acc_u128: U128 = acc.into();
+    let a_u128: U128 = a.into();
+    let b_u128: U128 = b.into();
+    let c_u128: U128 = c.into();
+
+    let res: U128 = a_u128 * b_u128 + acc_u128 + c_u128;
+    res.into()
+}
+
+impl core::ops::Multiply for BigUint {
+    fn multiply(self, other: Self) -> Self {
+        // Determine the length and setup variables
+        let self_number_of_limbs = self.limbs.len();
+        let other_number_of_limbs = other.limbs.len();
+        let mut result = Vec::with_capacity(self_number_of_limbs + other_number_of_limbs);
+        result.resize(self_number_of_limbs + other_number_of_limbs, 0);
+        let mut i = 0;
+
+        while i < self_number_of_limbs {
+            let mut j = 0;
+            let mut carry = 0;
+
+            while j < other_number_of_limbs {
+                // Multiply limbs and add to the result
+                let (c, res) = mac_with_carry_u64(
+                    result
+                        .get(i + j)
+                        .unwrap(),
+                    self.limbs
+                        .get(i)
+                        .unwrap(),
+                    other.limbs
+                        .get(j)
+                        .unwrap(),
+                    carry,
+                );
+
+                result.set(i + j, res);
+                carry = c;
+
+                j += 1;
+            }
+
+            // Propagate the carry to the next position
+            if carry > 0 {
+                result.set(
+                    i + other_number_of_limbs,
+                    result
+                        .get(i + other_number_of_limbs)
+                        .unwrap() + carry,
+                );
+            }
+
+            i += 1;
+        }
+
+        // Remove leading zeros
+        while result.len() > 1 && result.get(result.len() - 1) == Some(0) {
+            let _ = result.pop();
+        }
+
+        Self { limbs: result }
+    }
+}
+
+fn sub_with_borrow_u64(a: u64, b: u64, c: u64) -> (u64, u64) {
+    let a_u128: U128 = a.into();
+    let b_u128: U128 = b.into();
+    let c_u128: U128 = c.into();
+
+    let _ = disable_panic_on_overflow();
+    let diff: U128 = a_u128 - b_u128 - c_u128;
+    enable_panic_on_overflow();
+
+    let (diff_msw, diff_lsw): (u64, u64) = diff.into();
+
+    let carry = if diff_msw != 0 { 1 } else { 0 };
+    (diff_lsw, carry)
+}
+
+impl core::ops::Subtract for BigUint {
+    fn subtract(self, other: Self) -> Self {
+        // No negative numbers
+        if self < other {
```

Comment:
What do you think about this being a require statement with a custom error?

Reply:
The compiler reverts the same way when you subtract something like a `u64`. `U128` also performs in the same way. Perhaps we change this in the future when we decide what behavior should be expected and all `revert(0)` cases are removed from the std-lib